ement.el

```diff
@@ -222,6 +216,51 @@ call `pop-to-buffer'."
 
 ;;;; Functions
 
+(defun ement--read-username-and-server (prompt &optional initial-input history
+                                               default-value
+                                               inherit-input-method)
+  "Read Matrix user ID, submitted in the form \"@username:server\".
+
+Return (user-input . plist) where plist is `auth-source-search'-compatible plist."
+  (cl-loop
+   as user-id = (read-string prompt
+                             initial-input history default-value
+                             inherit-input-method)
+   if (not (string-match
+            (rx bos "@" (group (1+ (not (any ":")))) ; Username
+                ":" (group (optional (1+ (not (any blank)))))) ; Server name
+            user-id))
+   do (minibuffer-message "Invalid user ID format: use @USERNAME:SERVER")
+   else return (cons user-id
+                     (list :user (match-string 1 user-id)
+                           :host (match-string 2 user-id)))))
+
+(defun ement--read-user-credentials (&optional initial-input history
+                                               default-value
+                                               inherit-input-method)
+  "Read Matrix user credentials.  Return list (user-id password).
+
+Optional arguments refer to reading user ID."
+  (cl-destructuring-bind (user-id . plist) (ement--read-username-and-server
+                                            "User ID: "
+                                            initial-input history
+                                            default-value
+                                            inherit-input-method)
+    (list user-id
+          (save-match-data
```

Comment:
Generally, I think having to use `save-match-data` is a "code smell."  We shouldn't be relying on the match data across function calls.  (Again, this probably won't matter if we combine these functions.)

Reply:
Yes.  I hope you're fine with `ement-connect` supporting more rich input.